src/librustdoc/passes/collect_intra_doc_links/early.rs

```diff
@@ -1,75 +1,148 @@
+                        .collect(),
+                );
+            }
+        }
+    }
+
+    /// Add traits in scope for links in impls collected by the `collect-intra-doc-links` pass.
+    /// That pass filters impls using type-based information, but we don't yet have such
+    /// information here, so we just conservatively calculate traits in scope for *all* modules
+    /// having impls in them.
+    fn add_foreign_traits_in_scope(&mut self) {
+        for cnum in self.resolver.cstore().crates_untracked() {
+            // For some reason rustdoc also wants to know traits in scope for every crate root.
+            self.add_traits_in_scope(cnum.as_def_id());
+
+            for impl_def_id in
+                self.resolver.cstore().all_trait_implementations_def_ids_untracked(cnum)
+            {
+                if let Some(module_id) = self.resolver.parent(impl_def_id) {
+                    self.add_traits_in_scope(module_id);
+                }
+            }
+        }
+    }
 
-        // FIXME: this probably needs to consider inlining
-        let attrs = crate::clean::Attributes::from_ast(attrs, None);
+    fn load_links_in_attrs(&mut self, attrs: &[ast::Attribute], span: Span) {
+        // FIXME: this needs to consider export inlining.
+        let attrs = clean::Attributes::from_ast(attrs, None);
         for (parent_module, doc) in attrs.collapsed_doc_value_by_module_level() {
-            debug!(?doc);
+            let module_id = parent_module.unwrap_or(self.current_mod.to_def_id());
+
+            self.add_traits_in_scope(module_id);
+
             for link in markdown_links(&doc.as_str()) {
-                debug!(?link.link);
                 let path_str = if let Some(Ok(x)) = preprocess_link(&link) {
                     x.path_str
                 } else {
                     continue;
                 };
-                self.resolver.borrow_mut().access(|resolver| {
-                    let _ = resolver.resolve_str_path_error(
-                        span,
-                        &path_str,
-                        TypeNS,
-                        parent_module.unwrap_or(self.current_mod.to_def_id()),
-                    );
-                });
+                let _ = self.resolver.resolve_str_path_error(
+                    span,
+                    &path_str,
+                    Namespace::TypeNS,
+                    module_id,
+                );
+            }
+        }
+    }
+
+    /// Exports are replaced with item which they refer to, those items may have links in their
+    /// doc comments, those links are resolved at the item definition site, so we need to know
+    /// traits in scope at that definition site.
+    fn process_exports(&mut self, module_id: DefId) {
+        if !self.visited_mods.insert(module_id) {
+            return; // avoid infinite recursion
+        }
+
+        for export in self.resolver.module_exports(module_id) {
+            if export.vis == Visibility::Public {
```

Comment:
Hmm, this uses syntactic visibility, right? Would it be better to check if it's reachable from other crates?

Also, I'm not sure this is correct when `--document-private-items` is passed - can you add a test case for that? It would be something like:
```rust
mod inner {
    trait Trait {}
    /// Link to [Trait]
    fn foo() {}
}

Reply:
This is probably the source of CI failures, I didn't consider the `--document-private-items` mode at all.

Reply:
I think it would be enough to just remove the `if vis == Public` check.

Reply:
The CI failure was due to a removed import, not rustdoc logic.
I've run the whole `x.py test --stage 2` with `compiler-docs = true`, which documents private items, and didn't encountered any panics.
So given that `fn build_module` doesn't inline reexports without `ty::Visibility::Public`, the current logic seems to be correct.

(The test snippet in https://github.com/rust-lang/rust/pull/88679#discussion_r703055115 is not relevant to this code because it doesn't contain associated items in links or reexports.)